pkg/karmadactl/util/check.go

```diff
@@ -6,72 +6,135 @@ import (
+			if err != nil {
+				return err
+			} else if status == StatusReady {
+				return nil
+			}
 		}
-		return true, nil
-	})
-	if pollError != nil {
-		return fmt.Errorf("wait for Statefulset(%s/%s) rollout: %v: %v", sts.GetNamespace(), sts.GetName(), pollError, lastErr)
 	}
-	return nil
 }
 
 // WaitForDeploymentRollout  wait for Deployment reaches the ready state or timeout.
-func WaitForDeploymentRollout(c kubernetes.Interface, dep *appsv1.Deployment, timeoutSeconds int) error {
-	var lastErr error
-	pollError := wait.PollImmediate(time.Second, time.Duration(timeoutSeconds)*time.Second, func() (bool, error) {
-		d, err := c.AppsV1().Deployments(dep.GetNamespace()).Get(context.TODO(), dep.GetName(), metav1.GetOptions{})
-		if err != nil {
-			lastErr = err
-			return false, nil
-		}
-		if d.Generation != d.Status.ObservedGeneration {
-			lastErr = fmt.Errorf("current generation %d, observed generation %d",
-				d.Generation, d.Status.ObservedGeneration)
-			return false, nil
-		}
-		if (d.Spec.Replicas != nil) && (d.Status.UpdatedReplicas < *d.Spec.Replicas) {
-			lastErr = fmt.Errorf("the number of pods targeted by the deployment (%d pods) is different "+
-				"from the number of pods targeted by the deployment that have the desired template spec (%d pods)",
-				*d.Spec.Replicas, d.Status.UpdatedReplicas)
-			return false, nil
+func WaitForDeploymentRollout(c kubernetes.Interface, dep *appsv1.Deployment, timeoutSeconds time.Duration) error {
+	depWatcher, err := c.AppsV1().Deployments(dep.GetNamespace()).Watch(context.Background(), metav1.SingleObject(dep.ObjectMeta))
+	if err != nil {
+		return err
+	}
+
+	for {
+		select {
+		case <-time.After(timeoutSeconds * time.Second):
+			return nil
+		case event, ok := <-depWatcher.ResultChan():
+			if !ok {
+				return fmt.Errorf("wait for Deployment(%s/%s) rollout: channel broken", dep.GetNamespace(), dep.GetName())
+			}
+
+			status, err := deploymentStatus(event.Object)
+			if err != nil {
+				return err
+			} else if status == StatusReady {
+				return nil
+			}
 		}
-		if d.Status.AvailableReplicas < d.Status.UpdatedReplicas {
-			lastErr = fmt.Errorf("expected %d replicas, got %d available replicas",
-				d.Status.UpdatedReplicas, d.Status.AvailableReplicas)
-			return false, nil
+	}
+}
+
+// Reference https://github.com/kubernetes-sigs/application/blob/e5329b1b083ece8abb4b4efaf738ab9277fbb2ce/controllers/status.go#L78-#L92
+func stsStatus(o runtime.Object) (string, error) {
+	u, err := runtime.DefaultUnstructuredConverter.ToUnstructured(o)
+	if err != nil {
+		return StatusUnknown, err
+	}
+
+	sts := &appsv1.StatefulSet{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u, sts); err != nil {
+		return StatusUnknown, err
+	}
+
+	if sts.Status.ObservedGeneration == sts.Generation &&
+		sts.Status.Replicas == *sts.Spec.Replicas &&
+		sts.Status.ReadyReplicas == *sts.Spec.Replicas &&
+		sts.Status.CurrentReplicas == *sts.Spec.Replicas {
```

Comment:
Yes, when you update the sts.Spec.Replicas field to a higher value, Kubernetes will start creating new replicas of the StatefulSet. This process can take some time depending on the size of the StatefulSet and the available resources in the cluster.

During this process, the sts.Status.CurrentReplicas field will gradually increase as each new replica is created and becomes ready. However, it may take some time for all replicas to become ready, so there may be a period of time where sts.Spec.Replicas is greater than sts.Status.CurrentReplicas.

Once all replicas have been created and become ready, sts.Status.CurrentReplicas should equal sts.Spec.Replicas. If there are any issues during the creation of new replicas, the sts.Status.CurrentReplicas field may be lower than sts.Spec.Replicas, indicating that there are some replicas that are not yet ready or have failed to start.

( However, your screenshot does not seem to be related to the problem.

Reply:
ok，i got it. Thanks fo your explanation.